File: MaxSubArray/maxsubarray.cpp
#include <iostream>
using namespace std;

const int MAX_INT = 0x7fffffff;
const int MIN_INT = 0x80000000;

int max(int a, int b, int c) {
    if (a > b && a > c) return a;
    else if(b > a && b > c) return b;
    else return c;
}

void showArray(int* arr, int len, const char ch='\n') {
    for (int i = 0; i < len; ++i)
        cout << arr[i] << ch;
    if (ch != '\n') cout << endl;
}

void findCrossSubArray(int *arr, int low, int mid, int high, int returnArr[]) {
    int leftSum = MIN_INT;
    int lsum = 0;
    int max_left, max_right;
    int i, j;
    for (i = mid; i >= 0; i--) {
        lsum = lsum + arr[i];
        if (lsum > leftSum){
            leftSum = lsum;
            max_left = i;
        }
    }

    int rightSum = MIN_INT;
    int rsum = 0;
    for (j = mid + 1; j <= high; ++j) {
        rsum = rsum + arr[j];
        if (rsum > rightSum){
            rightSum = rsum;
            max_right = j;
        }
    }
    returnArr[0] = max_left; returnArr[1] = max_right; returnArr[2] = leftSum + rightSum;
}

void findMaxSubArray(int *arr, int low, int high, int resultArr[]) {
    // cout << "low is "<< low << "and high is" << high << endl;
    if(low == high) {
        resultArr[0] = low;
        resultArr[1] = high;
        resultArr[2] = arr[low];
        return;
    } else {
        int lResArr[3], rResArr[3], cResArr[3];
        int mid = (high + low) / 2;
        findCrossSubArray(arr, low, mid, high, cResArr);
        findMaxSubArray(arr, low, mid, lResArr);
        findMaxSubArray(arr, mid + 1, high, rResArr);
        // cout << cResArr[0] <<"--" << cResArr[1] << "--" << cResArr[2] << endl;
        if ((max(lResArr[2], rResArr[2], cResArr[2])) == lResArr[2]){
            resultArr[0] = lResArr[0];
            resultArr[1] = lResArr[1];
            resultArr[2] = lResArr[2];
        }
        else if ((max(lResArr[2], rResArr[2], cResArr[2])) == rResArr[2]){
            resultArr[0] = rResArr[0];
            resultArr[1] = rResArr[1];
            resultArr[2] = rResArr[2];
        }
        else {
            resultArr[0] = cResArr[0];
            resultArr[1] = cResArr[1];
            resultArr[2] = cResArr[2];
        }
    }

}

int main () {
    cout << "The max is" << MAX_INT << "and the min is" << MIN_INT << endl;
    int k;
    cout << "请输入一个正整数k:";
    cin >> k;
    int *arr = new int[k];
    for(int i = 0; i < k; i++) {
        cin >> arr[i];
    }
    showArray(arr, k);
    int res[3];
    int resultArr[3] = {};
    findMaxSubArray(arr,0, k - 1, resultArr);
    // findCrossSubArray(arr,0,k/2,k, res);
    cout << "the resule is:" << resultArr[2] << endl;

    delete[] arr;
    return 0;
}

File: MaxSubArray/MaxSubArray.py
#!/usr/bin/python
#coding:utf-8

def FIND_MAX_CROSSING_SUBARRAY(A, low, mid, high):
    left_Sum = -float("inf")
    Sum = 0
    for i in range(mid, low - 1, -1):
        Sum = Sum + A[i]
        if Sum > left_Sum:
            left_Sum = Sum
            max_left = i

    right_Sum = -float("inf")
    Sum = 0
    max_right = 0
    for j in range(mid + 1, high + 1):
        Sum = Sum + A[j]
        if Sum > right_Sum:
            right_Sum = Sum
            max_right = j
    return max_left, max_right, left_Sum + right_Sum

def FIND_MAXIMUM_SUBARRAY(A, low, high):
    if low == high:
        return (low, high, A[low])
    else:
        mid = int((low + high) / 2)
        left_low, left_high, left_Sum = \
            FIND_MAXIMUM_SUBARRAY(A, low, mid)
        right_low, right_high, right_Sum = \
            FIND_MAXIMUM_SUBARRAY(A, mid + 1, high)
        cross_low, cross_high, cross_Sum = \
            FIND_MAX_CROSSING_SUBARRAY(A, low, mid, high)

        if max([left_Sum, right_Sum, cross_Sum]) == left_Sum:
            return left_low, left_high, left_Sum
        elif max([left_Sum, right_Sum, cross_Sum]) == right_Sum:
            return right_low, right_high, right_Sum
        else:
            return cross_low, cross_high, cross_Sum

if __name__ == '__main__':
    K = input()
    inputLine = raw_input()
    A = [int(a) for a in inputLine.split(' ')]
    A = A[:K]

    res = FIND_MAXIMUM_SUBARRAY(A, 0, K-1)
    if res[2] < 0:
        print 0, A[0], A[K-1]
    else:
        print res[2], A[res[0]], A[res[1]]
